### PR TITLE
fix(solana): listen on wallet's accountChanged

### PR DESCRIPTION
**Pull request by UncleBill**
closes #MF-77

### CODE REVIEW

File: packages/injected-script/main/GlobalVariableBridge/index.ts [hunk 1]
```diff
@@ -27,8 +27,8 @@ export function callRequest(path: string, id: number, request: unknown) {
     handlePromise(id, () => read(path).request(request))
 }
 
-export function execute(path: string, id: number) {
-    handlePromise(id, () => read(path)())
+export function execute(path: string, id: number, ...args: any) {
```
Comment (UncleBill) [inline packages/injected-script/main/GlobalVariableBridge/index.ts]:
```suggestion
export function execute(path: string, id: number, ...args: any[]) {
```

File: packages/injected-script/main/GlobalVariableBridge/index.ts [hunk 2]
```diff
@@ -27,8 +27,8 @@ export function callRequest(path: string, id: number, request: unknown) {
     handlePromise(id, () => read(path).request(request))
 }
 
-export function execute(path: string, id: number) {
-    handlePromise(id, () => read(path)())
+export function execute(path: string, id: number, ...args: any[]) {
```
Comment (Lanttcat) [inline packages/injected-script/main/GlobalVariableBridge/index.ts]: If an explicit type cannot be given, then use unknow?

File: packages/plugins/Solana/src/wallet/connect.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+type BNLike = {
+    _bn: string
+}
+
+export async function storeConnection(pubKey: string | BNLike, chainId?: number) {
+    const base58Key = typeof pubKey === 'string' ? pubKey : hexToBase58(pubKey._bn)
+    const storage = getStorage()
+    await storage.publicKey.setValue(base58Key)
+    if (chainId) {
+        await storage.network.setValue(chainId)
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/wallet/connect.ts]:
```suggestion
        await storage.chainId.setValue(chainId)
```

File: packages/plugins/Solana/src/storage/index.ts [hunk 4]
```diff
@@ -14,3 +15,15 @@ export function setupStorage(_: typeof storage) {
...
+
+type BNLike = {
+    _bn: string
+}
+export async function storeConnection(pubKey: string | BNLike, chainId?: number) {
+    const base58Key = typeof pubKey === 'string' ? pubKey : hexToBase58(pubKey._bn)
+    const storage = getStorage()
+    await storage.publicKey.setValue(base58Key)
+    if (chainId) {
+        await storage.network.setValue(chainId)
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/storage/index.ts]:
```suggestion
        await storage.chainId.setValue(chainId)
```

File: packages/plugins/Solana/src/wallet/connect.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    }
+}
+export async function connectWallet(init = false) {
+    const rsp = await bridgedSolanaProvider.connect({ onlyIfTrusted: init })
+    if (rsp?.publicKey) {
+        await storeConnection(rsp.publicKey)
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/wallet/connect.ts:26]: Why it's not necessary also to store `chainId`?

File: packages/plugins/Solana/src/wallet/connect.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    return rsp?.publicKey
+}
+
+export async function watchAccount() {
+    bridgedSolanaProvider.on('accountChanged', async (publicKey) => {
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/wallet/connect.ts:36]: So what about `chainChanged` event? Does any event exist for a similar purpose on Phantom? Also `disconnect` event.

File: packages/plugins/Solana/src/wallet/connect.ts (hunk 5 above)
Comment (guanbinrui) [reply packages/plugins/Solana/src/wallet/connect.ts:26]: It depends on implementation. Phantom doesn't allow to retrieve `chainId` (The current network setting).

File: packages/plugins/Solana/src/wallet/connect.ts (hunk 6 above)
Comment (UncleBill) [reply packages/plugins/Solana/src/wallet/connect.ts:36]: > So what about chainChanged event

Phantom doesn't provide such event.

File: packages/plugins/Solana/src/storage/index.ts [hunk 7]
```diff
@@ -14,3 +15,15 @@ export function setupStorage(_: typeof storage) {
 export function getStorage() {
     return storage.storage
 }
+
+type BNLike = {
+    _bn: string
+}
+export async function storeConnection(pubKey: string | BNLike, chainId?: number) {
```
Comment (guanbinrui) [inline packages/plugins/Solana/src/storage/index.ts]: Solana should have an enumerate value for chain IDs. Therefore, you could change the `number` type into `ChainId`.

File: packages/plugins/Solana/src/wallet/connect.ts (hunk 6 above)
Comment (guanbinrui) [reply packages/plugins/Solana/src/wallet/connect.ts:36]: Okay, but maybe we should not suppose the provider is Phantom.

File: packages/plugins/Solana/src/storage/index.ts (hunk 7 above)
Comment (UncleBill) [reply packages/plugins/Solana/src/storage/index.ts]: Fixed in https://github.com/DimensionDev/Maskbook/pull/5921/commits/e6987f670b158fa0f4b121000a6fba537bbda906

**Review by guanbinrui — APPROVED** [review]
LGTM @DimensionDev/qa

**Comment by AinuX** [comment]
Checked

**Pull request merged**